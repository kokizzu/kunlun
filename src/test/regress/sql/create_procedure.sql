CALL nonexistent();  -- error
CALL random();  -- error
--DDL_STATEMENT_BEGIN--
CREATE FUNCTION cp_testfunc1(a int) RETURNS int LANGUAGE SQL AS $$ SELECT a $$;
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
CREATE TABLE cp_test (a int, b text);
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
CREATE PROCEDURE ptest1(x text)
LANGUAGE SQL
AS $$
INSERT INTO cp_test VALUES (1, x);
$$;
--DDL_STATEMENT_END--

\df ptest1
SELECT pg_get_functiondef('ptest1'::regproc);

-- show only normal functions
\dfn public.*test*1

-- show only procedures
\dfp public.*test*1

SELECT ptest1('x');  -- error
CALL ptest1('a');  -- ok
CALL ptest1('xy' || 'zzy');  -- ok, constant-folded arg
CALL ptest1(substring(random()::numeric(20,15)::text, 1, 1));  -- ok, volatile arg

SELECT * FROM cp_test ORDER BY b COLLATE "C";


--DDL_STATEMENT_BEGIN--
CREATE PROCEDURE ptest2()
LANGUAGE SQL
AS $$
SELECT 5;
$$;
--DDL_STATEMENT_END--

CALL ptest2();


-- nested CALL
delete from cp_test;

--DDL_STATEMENT_BEGIN--
CREATE PROCEDURE ptest3(y text)
LANGUAGE SQL
AS $$
CALL ptest1(y);
CALL ptest1($1);
$$;
--DDL_STATEMENT_END--

CALL ptest3('b');

SELECT * FROM cp_test;


-- output arguments

--DDL_STATEMENT_BEGIN--
CREATE PROCEDURE ptest4a(INOUT a int, INOUT b int)
LANGUAGE SQL
AS $$
SELECT 1, 2;
$$;
--DDL_STATEMENT_END--

CALL ptest4a(NULL, NULL);

--DDL_STATEMENT_BEGIN--
CREATE PROCEDURE ptest4b(INOUT b int, INOUT a int)
LANGUAGE SQL
AS $$
CALL ptest4a(a, b);  -- error, not supported
$$;
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
DROP PROCEDURE ptest4a;
--DDL_STATEMENT_END--


-- named and default parameters

--DDL_STATEMENT_BEGIN--
CREATE OR REPLACE PROCEDURE ptest5(a int, b text, c int default 100)
LANGUAGE SQL
AS $$
INSERT INTO cp_test VALUES(a, b);
INSERT INTO cp_test VALUES(c, b);
$$;
--DDL_STATEMENT_END--

delete from cp_test;

CALL ptest5(10, 'Hello', 20);
CALL ptest5(10, 'Hello');
CALL ptest5(10, b => 'Hello');
CALL ptest5(b => 'Hello', a => 10);

SELECT * FROM cp_test;


-- polymorphic types

--DDL_STATEMENT_BEGIN--
CREATE PROCEDURE ptest6(a int, b anyelement)
LANGUAGE SQL
AS $$
SELECT NULL::int;
$$;
--DDL_STATEMENT_END--

CALL ptest6(1, 2);


-- collation assignment

--DDL_STATEMENT_BEGIN--
CREATE PROCEDURE ptest7(a text, b text)
LANGUAGE SQL
AS $$
SELECT a = b;
$$;
--DDL_STATEMENT_END--

CALL ptest7(least('a', 'b'), 'a');


-- various error cases

CALL version();  -- error: not a procedure
CALL sum(1);  -- error: not a procedure

--DDL_STATEMENT_BEGIN--
CREATE PROCEDURE ptestx() LANGUAGE SQL WINDOW AS $$ INSERT INTO cp_test VALUES (1, 'a') $$;
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
CREATE PROCEDURE ptestx() LANGUAGE SQL STRICT AS $$ INSERT INTO cp_test VALUES (1, 'a') $$;
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
CREATE PROCEDURE ptestx(OUT a int) LANGUAGE SQL AS $$ INSERT INTO cp_test VALUES (1, 'a') $$;
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
ALTER PROCEDURE ptest1(text) STRICT;
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
ALTER FUNCTION ptest1(text) VOLATILE;  -- error: not a function
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
ALTER PROCEDURE cp_testfunc1(int) VOLATILE;  -- error: not a procedure
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
ALTER PROCEDURE nonexistent() VOLATILE;
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
DROP FUNCTION ptest1(text);  -- error: not a function
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
DROP PROCEDURE cp_testfunc1(int);  -- error: not a procedure
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
DROP PROCEDURE nonexistent();
--DDL_STATEMENT_END--


-- privileges

--DDL_STATEMENT_BEGIN--
CREATE USER regress_cp_user1;
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
GRANT INSERT ON cp_test TO regress_cp_user1;
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
REVOKE EXECUTE ON PROCEDURE ptest1(text) FROM PUBLIC;
--DDL_STATEMENT_END--
SET ROLE regress_cp_user1;
CALL ptest1('a');  -- error
RESET ROLE;
--DDL_STATEMENT_BEGIN--
GRANT EXECUTE ON PROCEDURE ptest1(text) TO regress_cp_user1;
--DDL_STATEMENT_END--
SET ROLE regress_cp_user1;
CALL ptest1('a');  -- ok
RESET ROLE;


-- ROUTINE syntax

--DDL_STATEMENT_BEGIN--
ALTER ROUTINE cp_testfunc1(int) RENAME TO cp_testfunc1a;
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
ALTER ROUTINE cp_testfunc1a RENAME TO cp_testfunc1;
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
ALTER ROUTINE ptest1(text) RENAME TO ptest1a;
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
ALTER ROUTINE ptest1a RENAME TO ptest1;
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
DROP ROUTINE cp_testfunc1(int);
--DDL_STATEMENT_END--

-- cleanup

--DDL_STATEMENT_BEGIN--
DROP PROCEDURE ptest1;
--DDL_STATEMENT_END--
--DDL_STATEMENT_BEGIN--
DROP PROCEDURE ptest2;
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
DROP TABLE cp_test;
--DDL_STATEMENT_END--

--DDL_STATEMENT_BEGIN--
DROP USER regress_cp_user1;
--DDL_STATEMENT_END--