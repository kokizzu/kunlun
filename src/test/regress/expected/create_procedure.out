CALL nonexistent();  -- error
psql:sql/create_procedure.sql:1: ERROR:  procedure nonexistent() does not exist
LINE 1: CALL nonexistent();
             ^
HINT:  No procedure matches the given name and argument types. You might need to add explicit type casts.
CALL random();  -- error
psql:sql/create_procedure.sql:2: ERROR:  random() is not a procedure
LINE 1: CALL random();
             ^
HINT:  To call a function, use SELECT.
CREATE FUNCTION cp_testfunc1(a int) RETURNS int LANGUAGE SQL AS $$ SELECT a $$;
CREATE FUNCTION
CREATE TABLE cp_test (a int, b text);
CREATE TABLE
CREATE PROCEDURE ptest1(x text)
LANGUAGE SQL
AS $$
INSERT INTO cp_test VALUES (1, x);
$$;
CREATE PROCEDURE
\df ptest1
                        List of functions
 Schema |  Name  | Result data type | Argument data types | Type 
--------+--------+------------------+---------------------+------
 public | ptest1 |                  | x text              | proc
(1 row)

SELECT pg_get_functiondef('ptest1'::regproc);
                pg_get_functiondef                 
---------------------------------------------------
 CREATE OR REPLACE PROCEDURE public.ptest1(x text)+
  LANGUAGE sql                                    +
 AS $procedure$                                   +
 INSERT INTO cp_test VALUES (1, x);               +
 $procedure$                                      +
 
(1 row)

-- show only normal functions
\dfn public.*test*1
                           List of functions
 Schema |     Name     | Result data type | Argument data types | Type 
--------+--------------+------------------+---------------------+------
 public | cp_testfunc1 | integer          | a integer           | func
(1 row)

-- show only procedures
\dfp public.*test*1
                        List of functions
 Schema |  Name  | Result data type | Argument data types | Type 
--------+--------+------------------+---------------------+------
 public | ptest1 |                  | x text              | proc
(1 row)

SELECT ptest1('x');  -- error
psql:sql/create_procedure.sql:22: ERROR:  ptest1(unknown) is a procedure
LINE 1: SELECT ptest1('x');
               ^
HINT:  To call a procedure, use CALL.
CALL ptest1('a');  -- ok
CALL
CALL ptest1('xy' || 'zzy');  -- ok, constant-folded arg
CALL
CALL ptest1(substring(random()::numeric(20,15)::text, 1, 1));  -- ok, volatile arg
CALL
SELECT * FROM cp_test ORDER BY b COLLATE "C";
 a |   b   
---+-------
 1 | 0
 1 | a
 1 | xyzzy
(3 rows)

CREATE PROCEDURE ptest2()
LANGUAGE SQL
AS $$
SELECT 5;
$$;
CREATE PROCEDURE
CALL ptest2();
CALL
-- nested CALL
TRUNCATE cp_test;
TRUNCATE TABLE
CREATE PROCEDURE ptest3(y text)
LANGUAGE SQL
AS $$
CALL ptest1(y);
CALL ptest1($1);
$$;
CREATE PROCEDURE
CALL ptest3('b');
CALL
SELECT * FROM cp_test;
 a | b 
---+---
 1 | b
 1 | b
(2 rows)

-- output arguments
CREATE PROCEDURE ptest4a(INOUT a int, INOUT b int)
LANGUAGE SQL
AS $$
SELECT 1, 2;
$$;
CREATE PROCEDURE
CALL ptest4a(NULL, NULL);
 a | b 
---+---
 1 | 2
(1 row)

CREATE PROCEDURE ptest4b(INOUT b int, INOUT a int)
LANGUAGE SQL
AS $$
CALL ptest4a(a, b);  -- error, not supported
$$;
psql:sql/create_procedure.sql:68: ERROR:  calling procedures with output arguments is not supported in SQL functions
CONTEXT:  SQL function "ptest4b"
DROP PROCEDURE ptest4a;
DROP PROCEDURE
-- named and default parameters
CREATE OR REPLACE PROCEDURE ptest5(a int, b text, c int default 100)
LANGUAGE SQL
AS $$
INSERT INTO cp_test VALUES(a, b);
INSERT INTO cp_test VALUES(c, b);
$$;
CREATE PROCEDURE
TRUNCATE cp_test;
TRUNCATE TABLE
CALL ptest5(10, 'Hello', 20);
CALL
CALL ptest5(10, 'Hello');
CALL
CALL ptest5(10, b => 'Hello');
CALL
CALL ptest5(b => 'Hello', a => 10);
CALL
SELECT * FROM cp_test;
  a  |   b   
-----+-------
  10 | Hello
  20 | Hello
  10 | Hello
 100 | Hello
  10 | Hello
 100 | Hello
  10 | Hello
 100 | Hello
(8 rows)

-- polymorphic types
CREATE PROCEDURE ptest6(a int, b anyelement)
LANGUAGE SQL
AS $$
SELECT NULL::int;
$$;
CREATE PROCEDURE
CALL ptest6(1, 2);
CALL
-- collation assignment
CREATE PROCEDURE ptest7(a text, b text)
LANGUAGE SQL
AS $$
SELECT a = b;
$$;
CREATE PROCEDURE
CALL ptest7(least('a', 'b'), 'a');
CALL
-- various error cases
CALL version();  -- error: not a procedure
psql:sql/create_procedure.sql:116: ERROR:  version() is not a procedure
LINE 1: CALL version();
             ^
HINT:  To call a function, use SELECT.
CALL sum(1);  -- error: not a procedure
psql:sql/create_procedure.sql:117: ERROR:  sum(integer) is not a procedure
LINE 1: CALL sum(1);
             ^
HINT:  To call a function, use SELECT.
CREATE PROCEDURE ptestx() LANGUAGE SQL WINDOW AS $$ INSERT INTO cp_test VALUES (1, 'a') $$;
psql:sql/create_procedure.sql:119: ERROR:  invalid attribute in procedure definition
LINE 1: CREATE PROCEDURE ptestx() LANGUAGE SQL WINDOW AS $$ INSERT I...
                                               ^
CREATE PROCEDURE ptestx() LANGUAGE SQL STRICT AS $$ INSERT INTO cp_test VALUES (1, 'a') $$;
psql:sql/create_procedure.sql:120: ERROR:  invalid attribute in procedure definition
LINE 1: CREATE PROCEDURE ptestx() LANGUAGE SQL STRICT AS $$ INSERT I...
                                               ^
CREATE PROCEDURE ptestx(OUT a int) LANGUAGE SQL AS $$ INSERT INTO cp_test VALUES (1, 'a') $$;
psql:sql/create_procedure.sql:121: ERROR:  procedures cannot have OUT arguments
HINT:  INOUT arguments are permitted.
ALTER PROCEDURE ptest1(text) STRICT;
psql:sql/create_procedure.sql:123: ERROR:  invalid attribute in procedure definition
LINE 1: ALTER PROCEDURE ptest1(text) STRICT;
                                     ^
ALTER FUNCTION ptest1(text) VOLATILE;  -- error: not a function
psql:sql/create_procedure.sql:124: ERROR:  ptest1(text) is not a function
ALTER PROCEDURE cp_testfunc1(int) VOLATILE;  -- error: not a procedure
psql:sql/create_procedure.sql:125: ERROR:  cp_testfunc1(integer) is not a procedure
ALTER PROCEDURE nonexistent() VOLATILE;
psql:sql/create_procedure.sql:126: ERROR:  procedure nonexistent() does not exist
DROP FUNCTION ptest1(text);  -- error: not a function
psql:sql/create_procedure.sql:128: ERROR:  ptest1(text) is not a function
DROP PROCEDURE cp_testfunc1(int);  -- error: not a procedure
psql:sql/create_procedure.sql:129: ERROR:  cp_testfunc1(integer) is not a procedure
DROP PROCEDURE nonexistent();
psql:sql/create_procedure.sql:130: ERROR:  procedure nonexistent() does not exist
-- privileges
CREATE USER regress_cp_user1;
CREATE ROLE
GRANT INSERT ON cp_test TO regress_cp_user1;
GRANT
REVOKE EXECUTE ON PROCEDURE ptest1(text) FROM PUBLIC;
REVOKE
SET ROLE regress_cp_user1;
SET
CALL ptest1('a');  -- error
psql:sql/create_procedure.sql:139: ERROR:  permission denied for procedure ptest1
RESET ROLE;
RESET
GRANT EXECUTE ON PROCEDURE ptest1(text) TO regress_cp_user1;
GRANT
SET ROLE regress_cp_user1;
SET
CALL ptest1('a');  -- ok
CALL
RESET ROLE;
RESET
-- ROUTINE syntax
ALTER ROUTINE cp_testfunc1(int) RENAME TO cp_testfunc1a;
ALTER ROUTINE
ALTER ROUTINE cp_testfunc1a RENAME TO cp_testfunc1;
ALTER ROUTINE
ALTER ROUTINE ptest1(text) RENAME TO ptest1a;
ALTER ROUTINE
ALTER ROUTINE ptest1a RENAME TO ptest1;
ALTER ROUTINE
DROP ROUTINE cp_testfunc1(int);
DROP ROUTINE
-- cleanup
DROP PROCEDURE ptest1;
DROP PROCEDURE
DROP PROCEDURE ptest2;
DROP PROCEDURE
DROP TABLE cp_test;
DROP TABLE
DROP USER regress_cp_user1;
DROP ROLE
